aip/cloud/25001.md

```diff
@@ -0,0 +1,66 @@
+---
+id: 25001
+state: draft
+created: TBD
+placement:
+  category: gcp
+---
+
+# Empty Objects
+
+Fields of type Message **must not** be empty. Empty messages do not have uniform
+
+## Alternatives
+
+The presence of empty Messages are sometimes used to indicate that a flag is enabled. Alternatives to empty Messages could include a single boolean field within the Message to signify enablement or an Enum rather than an empty Message.
+
+### Add Field to Message
+For example:
+```proto
+message EnableMyFeature {
+}
+```
+Could be modeled instead as:
+
+```proto
+message MyFeature {
+  bool enabled = 1;
+}
+```
+
+### Move Option to Top-Level Enum
+
+For a situation where the presence of an empty Message indicates a certain option is being chosen, an alternative is to choose the option via an enum rather than the presence of a particular object.
+
+For example:
+```proto
+oneof access_model {
+  AccessModelFree free = 1
+  AccessModelQuota quota = 2
+}
+
+message AccessModelFree {
+}
+
+message AccessModelQuota {
+  int32 requests_per_day = 1;
+}
+```
+
+Could be modeled instead as:
+```proto
+enum AccessModel {
+  UNSPECIFIED = 0;
+  FREE = 1;
+  QUOTA = 1;
+}
+
+message AccessModelQuota {
+  int32 requests_per_day = 1;
+}
+```
```

Comment:
Except you've then got far more potential for conflicting options having their details specified at the same time.

The oneof pattern is much more sensible IMO... if tools are unable to handle empty messages, I'd far prefer them to be improved.

Reply:
I tend to agree that the proposal of sidelined enum configuration over the values in the enum itself has poor usability: it requires the user to look at two different values (the enum and the corresponding configuration message) to reason about which value is correct. 

I can see the argument about an empty top-level message, but what's the issue with an empty enum value? it's clear what the type of the object is from the enum itself.